OneSignal/OneSignal-iOS-SDK
# [5.0.0] Push subscription related fixes (3)

# Description
## One Line Summary
Three fixes that are push subscription related:
1. Send push subscription changes that were not captured before such as `app_version` or `sdk`
2. Initialize push subscriptions with a notification type
3. Send create push subscription request if we detect it has been deleted

## Details

### Motivation

**1. Check for push subscription updates on new sessions**
* Properties on the push subscription can change between app opens such as `app_version`, `sdk` version, or `device_os`
* These properties aren't driven by changes in the SDK and previously wouldn't be detected as changed
* Now, on each new session, we will ask the push subscription model to check itself for updates and enqueue those changes, if any.
* `app_version` particularly is important to keep updated for app developers

**2. Initialize the push subscription model with a `notification_type`**
- Create the push sub model with the `notification_types` at time of creation
- This is so that on the first app open of a new install, we can receive push prompt IAMs (the server expects to have an appropriate` notification_type` to determine eligibility)
- Typically, the notification_type sent up on User Create is `ERROR_PUSH_DELEGATE_NEVER_FIRED` or `-14`

**3. Recreate a push subscription to the server when we detect it is deleted**
* On new sessions, we fetch the SDK's user to update our local models of any changes to the user. If we detect our push subscription missing from the fetched user (indicating the push subscription may be deleted), we will send a create subscription request for our device's push subscription.

**4. On the user manager, add a convenience accessor to the push subscription model**
* We access the push subscription model via the model store instead of via `user.pushSubscriptionModel`
* If privacy consent is set in a wrong order, we may have sent requests, but by the time the request returns, we don't want to hydrate on a mock user's push subscription.
* So, we want to set important information such as tokens and subscription ID on the actual push subscription model that the SDK will rely on, but `user.pushSubscriptionModel` may operate on a mock push subscription.


# Testing
## Unit testing
No

## Manual testing
Tested on physical iPhone 13 on iOS 16.6

**Scenario 1: Send push subscription changes that were not captured before such as `app_version` or `sdk`**
- Install app with an `app_version` and `sdk` version
- Update `app_version` and `sdk` version and reopen app
- See that the new `app_version` and `sdk` version is sent to server

**Scenario 2: Initialize push subscriptions with a notification type**

- Prior to this PR, on the first app open, we will not receive the push prompt IAM. We will receive it on a second session. 
- Now, we will receive the push prompt IAM on the first app open. 


**Scenario 3: Send create push subscription request if we detect it has been deleted**
- In between 2 sessions, delete the push subscription via REST API
- Reopen app and see a new create subscription is sent
- Receive a new `subscription_id`

# Affected code checklist

# Checklist
## Overview

## Testing

## Final pass


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/OneSignal/OneSignal-iOS-SDK/1289)
